pkg/pillar/cmd/zedagent/zedagent.go

```diff
@@ -325,15 +328,18 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
 	zedagentCtx.fatalFlag = *zedagentCtx.fatalPtr
 
 	flowlogQueue := make(chan *flowlog.FlowMessage, flowlogQueueCap)
+	// Channels with buffer of size 1 are used with non-blocking send.
 	triggerDeviceInfo := make(chan destinationBitset, 1)
 	triggerHwInfo := make(chan destinationBitset, 1)
 	triggerLocationInfo := make(chan destinationBitset, 1)
-	triggerObjectInfo := make(chan infoForObjectKey, 1)
+	triggerObjectInfo := make(chan infoForObjectKey, 256)
```

Comment:
The goal is to never block so I added a new commit to this PR to make `triggerPublishObjectInfo*` behave just like [triggerPublishDevInfoToDest](https://github.com/milan-zededa/eve/blob/async-send/pkg/pillar/cmd/zedagent/zedagent.go#L1876-L1890) - i.e. submit trigger by sending message into the channel without blocking. For device info having buffer of size one is sufficient because there is only one device. Since objects can be many, I had to come up with some >1 buffer size. It is true that 256 might not be enough if device has rebooted and starts processing large config with many apps, volumes, network instances, etc. The simplest solution, currently implemented in this PR, is to start dropping when the queue is full (it is also quite likely that trigger for info publishing for a given object is already buffered). Alternatively, we can use a thread-safe queue with unlimited size for these triggers and use channel of size one to merely wake up the publishing task.
But when the currently-used fixed-sized buffered channel becomes full, it likely means that publishing is very slow or just keeps timing out due to not working connectivity. In such case I would be against triggering publishing for all existing objects as it may lead to a very long queue of info messages to publish when the connectivity is restored, thus operating with a long delay.

Reply:
I don't have a clear picture in my head, but isn't there two levels of channels? One is triggerObjectInfo which causes some code to run to generate a protobuf message, and then there is the deferred (kind of misnamed now but let's ignore that) processing by a different goroutine. The deferred can take forever if the network is flaky (as it manages to send some messages more messages might be added from the object info layer).

But shouldn't the object info layer be relatively fast? It might need to generate dozens or more messages but it shouldn't be subject to network timeouts with these diffs AFAIU (I didn't check where the LPS messages are sent though.)

Just trying to understand - with your conditional send the code is at least safe.

Reply:
It is true that a fair bit of work will end up on the `handleDeferredTask` plate, especially during peak times, such as after a reboot. However, `Publish*InfoToZedCloud` methods run by `objectInfoTask` try to process at least one (and the highest-priority) task synchronously: https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/zedagent/zedagent.go#L281-L283
@rouming Is that assessment correct?

Reply:
@milan-zededa yep, this is correct.

Reply:
@rouming why do we want to do that inline as opposed to sending all info messages to the controller from the separate goroutine? The comment for queueInfoToDest() talks about errors being ignored, which might be a red herring since we don't remove items from the deferred queue until the send has succeeded.

It would seem a lot more predictable to not have queueInfoToDest() sometimes take 2 minutes (or 6?) due to this and instead have the separate deferred goroutine deal with such delays.

Reply:
@eriknordmark I can only guess why the original behavior is to keep sending at least one request: this is because the deferred queue is not limited to any capacity and sender should experience back pressure, otherwise we end up having a queue growing if network is slow, but frequency of adding requests is high. This is only my speculations, which may make sense in corner cases (slow network, etc). Once I did refactoring if the deferred queue, i kept the original behavior as is. 

Errors are ignored only for the periodic queue, which is used for sending metrics info to LOC, those messages can be safely dropped, because of the recurring nature of the sender.

Reply:
That's not the reason and given these dekat issues we've discovered we must need to get rid of that inline send to avoid having unpredictable timeing behavior.

The code has transformed a lot but originally the deferred queue was only inspected every few minutes, hence all code sent and if that failed, add to deferred queue. That change to 1) add one message to queue 2) send one highest priority message, to make sure we make timely progress *and* send in priority order.
But the limit on the length of the queue was and is still at most one INFO message for each object in the system (device, app instance, volume instances, blobs, etc). 

So unless you see where things would break without that inline send, let's remove it so we don't have to revisit some other delay issue here in the future.